quelpa.el

```diff
@@ -1779,15 +1818,34 @@ the `quelpa' command has been run in the current Emacs session."
       (when quelpa-self-upgrade-p
         (quelpa-self-upgrade))
       (setq quelpa-cache
-            (cl-remove-if (lambda (package)
-                            (or (eq package 'quelpa)
-                                (not (package-installed-p package))))
-                          quelpa-cache :key #'car))
+            (cl-remove-if-not (lambda (item) (package-installed-p (car item))) quelpa-cache))
       (mapc (lambda (item)
               (when (package-installed-p (car (quelpa-arg-rcp item)))
                 (quelpa item)))
             quelpa-cache))))
 
+;;;###autoload
+(defun quelpa-upgrade (rcp)
+  "Upgrade a package found in `quelpa-cache'.
+RCP is a melpa recipe (list).
+\\[universal-argument] \\[quelpa-upgrade] for forced upgrade.
+\\[universal-argument] \\[universal-argument] \\[quelpa-upgrade] for upgrade using local tip."
+  (interactive
+   (list (when (quelpa-setup-p)
+           (let* ((quelpa-melpa-recipe-stores `(,quelpa-cache))
+                  (name (quelpa-interactive-candidate)))
+             (assoc name quelpa-cache)))))
+  (when rcp
+    (let* ((quelpa-upgrade-p t)
+           (action (prefix-numeric-value current-prefix-arg))
```

Comment:
It would be more idiomatic and easier to understand if the prefix arguments were handled in the `interactive` form by setting additional arguments to the function.  The idea of the `interactive` form is to contain all the interactive argument handling.  :)